rafthttp: use MergeLogger for rafthttp logging

yichengq opened the pull request:
rafthttp logs repeated messages whenever the same connection error
happens, and it easily becomes log spamming.
Use MergeLogger as the default logger to merge repeated log lines.

This performs the same functionality as logging control in peerStatus.
So I remove the logging control in peerStatus and only use MergeLogger.

A leader logs of 3-member cluster like this:
```
[first follower is online]
2015-11-02 13:04:10.077340 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream Message (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 319 repeated lines in 32.535044736s]
2015-11-02 13:04:10.077355 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream MsgApp v2 (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 319 repeated lines in 32.53498537s]
2015-11-02 13:04:10.077363 E | rafthttp: failed to dial fd422379fda50e48 on stream Message (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 320 repeated lines in 32.531640475s]
2015-11-02 13:04:10.077369 E | rafthttp: failed to dial fd422379fda50e48 on stream MsgApp v2 (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 320 repeated lines in 32.531619957s]
2015-11-02 13:04:10.077372 E | rafthttp: failed to write fd422379fda50e48 on pipeline (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 24 repeated lines in 31.937195869s]
2015-11-02 13:04:10.077375 E | rafthttp: failed to write 91bc3c398fb3c146 on pipeline (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 24 repeated lines in 31.937131233s]
2015-11-02 13:04:10.077418 I | rafthttp: the connection with 91bc3c398fb3c146 became active
...
[second follower is online]
2015-11-02 13:04:17.002485 E | rafthttp: failed to dial fd422379fda50e48 on stream MsgApp v2 (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 67 repeated lines in 6.867907047s]
2015-11-02 13:04:17.002502 E | rafthttp: failed to dial fd422379fda50e48 on stream Message (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 67 repeated lines in 6.867877413s]
2015-11-02 13:04:17.002508 E | rafthttp: failed to write fd422379fda50e48 on pipeline (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 80 repeated lines in 6.163153163s]
2015-11-02 13:04:17.002516 I | rafthttp: the connection with fd422379fda50e48 became active
```

A slow-snapshot-recovery follower acts like this:
```
2015-11-02 13:04:27.066564 W | rafthttp: dropped MsgAppResp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-02 13:04:27.066595 W | rafthttp: dropped MsgHeartbeatResp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-02 13:04:27.066870 W | rafthttp: dropped MsgProp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-02 13:04:28.956096 W | rafthttp: dropped MsgAppResp to 8211f1d0f64f3269 since pipeline's sending buffer is full [merged 20 repeated lines in 1.889445223s]
2015-11-02 13:04:28.956127 W | rafthttp: dropped MsgHeartbeatResp to 8211f1d0f64f3269 since pipeline's sending buffer is full [merged 46 repeated lines in 1.889413195s]
```

fixes #3462 in a general way

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,170 @@
+
+// MergeLogger merges repeated log lines and prints summary log lines.
+// It doesn't merge log at DEBUG and TRACE level because they should
+// be verbosal.
```
be verbose

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,170 @@
+	l.statusm[ln] = status
+}
+
+func (l *MergeLogger) printAllLoop() {
```
outputLoop?

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,170 @@
+// MergeLogger prints out summary log line for all log lines every period.
+type MergeLogger struct {
+	*capnslog.PackageLogger
+	Gap    time.Duration
```
add comments for public fields? Gap and Period?

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,170 @@
+// be verbosal.
+//
+// For one log line, MergeLogger prints out the line when the line appears
+// at the first time. Then it holds and merges lines with the same
```
The mergeLogger holds the same log line printed within the same `Gap`. For example...

yichengq commented:
Based on discussion OOL, I makes the mergeLogger simpler to have `mergePeriod` one concept only. And let mergeLogger fit into the various use cases. PTAL.

jonboulle commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,188 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutil
```
// Package logutil ....

jonboulle commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func print(logger *capnslog.PackageLogger, ln line, status *status) bool {
+	status.mu.Lock()
```
You have a per-status lock, but AFAICT you are only ever touching status while holding the global MergeLogger mutex - am I wrong?

xiang90 replied:
The for loop actually touches the protected map.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,188 @@
+
+func (l *MergeLogger) outputLoop() {
+	for now := range time.Tick(time.Second) {
+		l.mu.Lock()
```
We really should not hold a global lock when doing io (there are io in print).

xiang90 replied:
We probably should copy out the item in the map first.

yichengq replied:
ok

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,188 @@
+		status.mu.Unlock()
+		return
+	}
+
```
release the global lock while doing io?

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,188 @@
+	return l
+}
+
+func (l *MergeLogger) Info(mergePeriod time.Duration, entries ...interface{}) {
```
I think we only have two use cases: 

1. merge forever until flush.
2. merge with a fixed interval.

I would prefer to have separate methods than making the period an addition argument.

Like:
```go
l.Info()
l.MergedInfo()
```

@jonboulle any idea?

xiang90 replied:
I think I am wrong. We do have a lot of different mergeIntervals. I will read more to see if they are really needed.

xiang90 replied:
@yichengq Can you explain why `requestErrMergeInterval=30s` is needed?

yichengq replied:
The main use cases that I target for is errVersionIncompatibility. If there is one member in the cluster has too low version, all other members will print out errVersionIncompatibility in floods. 30s is a balance value for this case.

xiang90 replied:
I do not really think that is a big issue at least for now? How about let us keep this as simple as possible for now?

xiang90 replied:
The rule will be simple. We print at most 1 same warning per second. We do not print out anything for an inactive peer until it is active again.

yichengq replied:
ok.

yichengq commented:
Have removed mergePeriod parameter to make it simple in current stage as xiang suggests. Also, always release the global lock before doing IO. PTAL

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/coreos/pkg/capnslog"
+)
+
+var normalPeriod = time.Second
```
what is normalPeriod? If this is a global variable, it needs to be specific. I do not know what does normal mean without any context.

yichengq replied:
It is the merge period length of normal operations. will change to `defaultPeriod`.

xiang90 replied:
defaultWHATPeriod

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+// CachedErrorf caches the error message.
+// The cached messages are printed out only when calling Flush.
+func (l *MergeLogger) CachedErrorf(format string, args ...interface{}) {
+	l.merge(0, line{
```
if we make 1 second a const, we should make 0 a const too.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+	count int        // number of lines from start field
+}
+
+func (s *status) canPrint(now time.Time) bool {
```
I feel we should change this to `isMerging`. First it makes the code logic focus on `merge`. Second the code block will be more readable like 

```go
if s.isMerging(){
    continue
}
``` 

instead of

```go
if !s.canPrint() {
    continue
}
```

yichengq replied:
:+1: sure

yichengq replied:
`isMerging` is a little confusing. It also imply that if `isMerging` is false, it should not merge further log lines, but this is not the case here. I will rename it to `inMergePeriod`.

xiang90 replied:
I still do not understand why it is confusing.

yichengq replied:
When a log line is printed, logger always finds out its status and increase its count. So even if `isMerging` is false, it still merges the line into the status. This sounds confusing.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+func (l *MergeLogger) outputLoop() {
+	for now := range time.Tick(time.Second) {
```
shall we use the defined const here instead of time.Second?

yichengq replied:
No. If it becomes a const, it could be `outputInterval`.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+			if !status.canPrint(now) {
+				continue
+			}
+			if status.count == 0 {
```
status.IsEmpty?

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+	return s.period > 0 && s.start.Add(s.period).Before(now)
+}
+
+func (s *status) getAndReset(now time.Time) string {
```
I feel this method is very dirty and ugly since it does two separate things.

yichengq replied:
i will split it.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+	str   string
+}
+
+func (l line) attach(s string) line {
```
why this is attach? Shoudnt it be append?

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+
+// CachedErrorf caches the error message.
+// The cached messages are printed out only when calling Flush.
+func (l *MergeLogger) CachedErrorf(format string, args ...interface{}) {
```
ErrorfCached?

yichengq replied:
ok

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+// Flush flushes summary log lines, and halts merging in progress.
+func (l *MergeLogger) Flush() {
```
It seems to be that we should have a function merging the logic in flush and the inner loop of output loop.

yichengq replied:
done.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+	}
+}
+
+type status struct {
```
can you add comment about status?

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+	period time.Duration
+
+	mu    sync.Mutex // protect fields below
+	start time.Time  // time to start merging
```
Is this the time to start merging (time in future)? I think this is the last merging starting time.

yichengq replied:
This is not a time in future. It is the start time of latest merge period.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,217 @@
+
+	mu    sync.Mutex // protect fields below
+	start time.Time  // time to start merging
+	count int        // number of lines from start field
```
number of merged lines from starting.

yichengq commented:
Addressed all comments. PTAL

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,218 @@
+
+var (
+	defaultMergePeriod   = time.Second
+	permanentMergePeriod time.Duration
```
`permanent` is not a good description for what you want this to achieve i believe.

yichengq replied:
use `forever` instead.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,218 @@
+	count int       // number of merged lines from starting
+}
+
+func (s *status) inMergePeriod(now time.Time) bool {
```
isInMergePeriod

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,218 @@
+	}
+}
+
+func (l *MergeLogger) output(now time.Time, flush bool) {
```
flush -> forceFlush

xiang90 commented:
Looks OK to me. Defer to @jonboulle

xiang90 commented:
@yichengq BTW, there is no test. So have you actually tested the code path manually?

yichengq commented:
@xiang90 Yes. And here is what it looks today:

A leader logs of 3-member cluster like this:
```
[first follower is online]
2015-11-02 13:04:10.077340 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream Message (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 319 repeated lines in 32.535044736s]
2015-11-02 13:04:10.077355 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream MsgApp v2 (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 319 repeated lines in 32.53498537s]
2015-11-02 13:04:10.077363 E | rafthttp: failed to dial fd422379fda50e48 on stream Message (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 320 repeated lines in 32.531640475s]
2015-11-02 13:04:10.077369 E | rafthttp: failed to dial fd422379fda50e48 on stream MsgApp v2 (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 320 repeated lines in 32.531619957s]
2015-11-02 13:04:10.077372 E | rafthttp: failed to write fd422379fda50e48 on pipeline (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 24 repeated lines in 31.937195869s]
2015-11-02 13:04:10.077375 E | rafthttp: failed to write 91bc3c398fb3c146 on pipeline (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 24 repeated lines in 31.937131233s]
2015-11-02 13:04:10.077418 I | rafthttp: the connection with 91bc3c398fb3c146 became active
2015-11-02 13:04:10.134550 E | rafthttp: failed to dial fd422379fda50e48 on stream MsgApp v2 (dial tcp 127.0.0.1:32380: getsockopt: connection refused)
2015-11-02 13:04:10.134579 E | rafthttp: failed to dial fd422379fda50e48 on stream Message (dial tcp 127.0.0.1:32380: getsockopt: connection refused)
...
[second follower is online]
2015-11-02 13:04:17.002485 E | rafthttp: failed to dial fd422379fda50e48 on stream MsgApp v2 (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 67 repeated lines in 6.867907047s]
2015-11-02 13:04:17.002502 E | rafthttp: failed to dial fd422379fda50e48 on stream Message (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 67 repeated lines in 6.867877413s]
2015-11-02 13:04:17.002508 E | rafthttp: failed to write fd422379fda50e48 on pipeline (dial tcp 127.0.0.1:32380: getsockopt: connection refused) [merged 80 repeated lines in 6.163153163s]
2015-11-02 13:04:17.002516 I | rafthttp: the connection with fd422379fda50e48 became active
```

A slow-snapshot-recovery follower acts like this:
```
2015-11-02 13:04:27.066564 W | rafthttp: dropped MsgAppResp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-02 13:04:27.066595 W | rafthttp: dropped MsgHeartbeatResp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-02 13:04:27.066870 W | rafthttp: dropped MsgProp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-02 13:04:28.956096 W | rafthttp: dropped MsgAppResp to 8211f1d0f64f3269 since pipeline's sending buffer is full [merged 20 repeated lines in 1.889445223s]
2015-11-02 13:04:28.956127 W | rafthttp: dropped MsgHeartbeatResp to 8211f1d0f64f3269 since pipeline's sending buffer is full [merged 46 repeated lines in 1.889413195s]
```

It fixes the log spamming in slow-snapshot-recovery because multiple lines are merged into one.

When one peer becomes active, it flushes all cached log lines out, including unrelated log lines. This is a little verbose, but it is clear.

xiang90 commented:
> When one peer becomes active, it flushes all cached log lines out, including unrelated log lines. This is a little verbose, but it is clear.

I do not really like this idea. :(

yichengq commented:
> I do not really like this idea. :(

will improve. Let me try to filter out the related log lines.

xiang90 commented:
When the first follower recovers, why does it print out the other's error too? I do not think it makes sense.

xiang90 commented:
What I was hoping is: do not print out any repeated error when we know the peer was dead.

yichengq commented:
@xiang90 Changed to print related log lines when some peer becomes active. It only flushes out the log lines that has the match peer ID.

The log does become much readable now. Thanks for pointing it out.

yichengq commented:
Original PR discards all status errors in debug level, which is incorrect. Print merged lines in DEBUG level, so people could get all of them when enabling debug logging.

I keep the behavior to print out summary log line when flush. One reason is that it brings some info to users that how many errors happened and how long this error exists. Another reason is that the logger is designed to merge logs and print summary.

xiang90 commented:
> I keep the behavior to print out summary log line when flush. One reason is that it brings some info to users that how many errors happened and how long this error exists. Another reason is that the logger is designed to merge logs and print summary.

I still dislike this a lot. 

@jonboulle ^

yichengq commented:
@jonboulle For context, it prints out like this now:
```
2015-11-02 16:10:13.340084 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream Message (dial tcp 127.0.0.1:22380: getsockopt: connection refused)
2015-11-02 16:10:13.341287 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream MsgApp v2 (dial tcp 127.0.0.1:22380: getsockopt: connection refused)
2015-11-02 16:10:13.937107 E | rafthttp: failed to write 91bc3c398fb3c146 on pipeline (dial tcp 127.0.0.1:22380: getsockopt: connection refused)
...
2015-11-02 16:10:19.530784 E | rafthttp: failed to write 91bc3c398fb3c146 on pipeline (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 4 repeated lines in 5.593332067s]
2015-11-02 16:10:19.530806 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream Message (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 60 repeated lines in 6.19031404s]
2015-11-02 16:10:19.530813 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream MsgApp v2 (dial tcp 127.0.0.1:22380: getsockopt: connection refused) [merged 60 repeated lines in 6.189107814s]
2015-11-02 16:10:19.530885 I | rafthttp: the connection with 91bc3c398fb3c146 became active
```

xiang and I are talking about 4-7th log lines. These log lines are not printed out before.

xiang90 commented:
@yichengq Also the API become messy. Flush takes an argument called `keyword`. This is so strange and it is unnecessary in my opinion.

xiang90 commented:
@yichengq What I hope to see is simply

```go
2015-11-02 16:10:13.340084 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream Message (dial tcp 127.0.0.1:22380: getsockopt: connection refused)
2015-11-02 16:10:13.341287 E | rafthttp: failed to dial 91bc3c398fb3c146 on stream MsgApp v2 (dial tcp 127.0.0.1:22380: getsockopt: connection refused)
2015-11-02 16:10:13.937107 E | rafthttp: failed to write 91bc3c398fb3c146 on pipeline (dial tcp 127.0.0.1:22380: getsockopt: connection refused)
2015-11-02 16:10:13.330885 I | rafthttp: the connection with 91bc3c398fb3c146 became inactive
...
2015-11-02 16:10:19.530885 I | rafthttp: the connection with 91bc3c398fb3c146 became active
```

We do not print any repeated error when we know the peer is inactive in non-debug mode.

xiang90 commented on pkg/logutil/merge_logger.go:
```diff
@@ -0,0 +1,233 @@
+		// release the lock before IO operation
+		l.mu.Unlock()
+		// print out the repeated line at DEBUG level
+		l.PackageLogger.Logf(capnslog.DEBUG, ln.str)
```
what is this?!?!

yichengq commented:
> Also the API become messy. Flush takes an argument called keyword. This is so strange and it is unnecessary in my opinion.

We don't want Flush to clear the log cache of still-inactive peers to avoid unnecessary logs at the first time, so keyword is needed.

I think printing merged lines in DEBUG level is acceptable because it is kind of debug logging. xiang is against it because it magically changes the log level of the log.

@jonboulle Could you give some insights around this PR?

xiang90 commented:
> We don't want Flush to clear the log cache of still-inactive peers to avoid unnecessary logs at the first time, so keyword is needed.

Actually no. You should not print them out at the first place...

jonboulle commented:
sorry guys, don't have time to dive in deep to this right now, but from my reading: I'm finding it hard to see the utility of repeating the logging lines or capturing how long they've been occurring. I'd prefer the much simpler event-driven ("X is offline", nothing, "X is online") approach. The aggregation stuff just feels way too complex for little gain..

xiang90 commented:
> I'd prefer the much simpler event-driven ("X is offline", nothing, "X is online") approach. The aggregation stuff just feels way too complex for little gain..

I agree. This is also the approach I suggested.

yichengq commented:
@jonboulle @xiang90 The merge logger is designed to merge repeated logs together and prints out summary line at some point. If we don't want to see summary logs, merge logger should not be used at these places IMO. But per previous discussion, we want to use the logger as a general one for rafthttp, and replace the functionality of logging mechanism in peerStatus. That makes me struggled.

After discussion OOL with xiang, it is not a requirement to use a general logger for all cases. My current thought is to let merge logger focus on the places that needs to be merged then. For logging mechanism in peerStatus, I don't find a good way to integrate it into merge logger. If you have good thoughts around integration, please let me know.

Thoughts?

jonboulle commented:
@yichengq can you help me understand a scenario where you definitely want to use the merge logger?

yichengq commented:
@jonboulle The #3462 issue. When the connection is active, it may send too many messages at the same time, and rafthttp will print one line for each dropped message. It is good to aggregate them to see how many messages are dropped.

yichengq commented:
@jonboulle xiang suggests that maybe it could be done by adjusting the log output in `deactivate`, and i am thinking about it.

yichengq commented:
Make the change to log deactivate better. Output looks like this now:

A leader logs of 3-member cluster like this:
```
2015-11-03 13:07:12.698832 E | rafthttp: failed to dial fd422379fda50e48 on stream MsgApp v2 (dial tcp 127.0.0.1:32380: getsockopt: connection refused)
2015-11-03 13:07:12.698853 I | rafthttp: the connection with fd422379fda50e48 became inactive
....
2015-11-03 13:07:19.828892 I | rafthttp: the connection with fd422379fda50e48 became active
```

A slow-snapshot-recovery follower acts like this:
```
2015-11-03 13:07:29.926853 W | rafthttp: dropped MsgHeartbeatResp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-03 13:07:29.926912 W | rafthttp: dropped MsgAppResp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-03 13:07:29.927124 W | rafthttp: dropped MsgProp to 8211f1d0f64f3269 since pipeline's sending buffer is full
2015-11-03 13:07:31.789354 W | rafthttp: dropped MsgHeartbeatResp to 8211f1d0f64f3269 since pipeline's sending buffer is full [merged 45 repeated lines in 1.862264185s]
2015-11-03 13:07:31.789416 W | rafthttp: dropped MsgAppResp to 8211f1d0f64f3269 since pipeline's sending buffer is full [merged 17 repeated lines in 1.862202598s]
```

jonboulle commented on rafthttp/peer_status.go:
```diff
@@ -49,25 +47,21 @@ func (s *peerStatus) activate() {
 	}
-	s.failureMap[failure] = reason
-	plog.Errorf(logline)
+	plog.Debugf(msg)
```
this should probably be in an else, otherwise msg will be logged twice when s is active?

yichengq replied:
It returns in `if` block when s is active, so this is not a problem

jonboulle replied:
gahhh, I missed that with all the red, sorry

jonboulle commented:
Cool, this looks much better to me now! I think minimising the use of mergeLogger to that particular instance makes more sense.

yichengq commented:
@jonboulle Updated. I explictly point out the logging that uses merge mechanism. PTAL

jonboulle commented:
Great. LGTM

yichengq commented:
@xiang90 makes a good point that normal users probably cannot understand what `MsgXXX` means and the expected behavior for it. I would open it in a new issue.

Going to merge this PR because it reaches its goal.
